### Build Completed

:x: This pull request has errors. They will need to be addressed before it can be accepted. See below for more details. Some links are also available below for further assistance in addressing those issues.

#### Build Commands

```sh
mvn -B -fn -DrunITs=true -f pdi-dataservice-client/pom.xml clean install && mvn -B -f pdi-dataservice-client/pom.xml site
```

#### Cleanup Commands
```sh
mvn -B -f pdi-dataservice-client/pom.xml build-helper:remove-project-artifact
```

#### Changed files

```console
  pdi-dataservice-client/src/main/java/org/pentaho/di/trans/dataservice/client/DataServiceConnectionInformation.java
  pdi-dataservice-client/src/test/java/org/pentaho/di/trans/dataservice/client/DataServiceConnectionInformationTest.java
```

Newly Broken Tests:
======================
###### org.pentaho.di.trans.dataservice.client.DataServiceConnectionInformationTest.testBadXmlLogsInfoAndReturnsLocalMeta:

:red_circle: `java.lang.NoClassDefFoundError`

```java
java.lang.NoClassDefFoundError: org/json/simple/JSONObject
	at org.pentaho.di.trans.dataservice.client.DataServiceConnectionInformationTest.expectRepository(DataServiceConnectionInformationTest.java:197)
	at org.pentaho.di.trans.dataservice.client.DataServiceConnectionInformationTest.testBadXmlLogsInfoAndReturnsLocalMeta(DataServiceConnectionInformationTest.java:180)
Caused by: java.lang.ClassNotFoundException: org.json.simple.JSONObject
	at org.pentaho.di.trans.dataservice.client.DataServiceCon
```
###### org.pentaho.di.trans.dataservice.client.DataServiceConnectionInformationTest.testDIRepositoriesReturnsMetaToSameHost:

:red_circle: `java.lang.NoClassDefFoundError`

```java
java.lang.NoClassDefFoundError: org/json/simple/JSONObject
	at org.pentaho.di.trans.dataservice.client.DataServiceConnectionInformationTest.expectRepository(DataServiceConnectionInformationTest.java:197)
	at org.pentaho.di.trans.dataservice.client.DataServiceConnectionInformationTest.testDIRepositoriesReturnsMetaToSameHost(DataServiceConnectionInformationTest.java:116)
Caused by: java.lang.ClassNotFoundException: org.json.simple.JSONObject
	at org.pentaho.di.trans.dataservice.client.DataServiceC
```
###### org.pentaho.di.trans.dataservice.client.DataServiceConnectionInformationTest.testOtherRepositoryTypesReturnLocalMeta:

:red_circle: `java.lang.NoClassDefFoundError`

```java
java.lang.NoClassDefFoundError: org/json/simple/JSONObject
	at org.pentaho.di.trans.dataservice.client.DataServiceConnectionInformationTest.expectRepository(DataServiceConnectionInformationTest.java:197)
	at org.pentaho.di.trans.dataservice.client.DataServiceConnectionInformationTest.testOtherRepositoryTypesReturnLocalMeta(DataServiceConnectionInformationTest.java:147)
Caused by: java.lang.ClassNotFoundException: org.json.simple.JSONObject
	at org.pentaho.di.trans.dataservice.client.DataServiceC
```
###### org.pentaho.di.trans.dataservice.client.DataServiceConnectionInformationTest.testWrongTagLogsInfoAndReturnsLocalMeta:

:red_circle: `java.lang.NoClassDefFoundError`

```java
java.lang.NoClassDefFoundError: org/json/simple/JSONObject
	at org.pentaho.di.trans.dataservice.client.DataServiceConnectionInformationTest.expectRepository(DataServiceConnectionInformationTest.java:197)
	at org.pentaho.di.trans.dataservice.client.DataServiceConnectionInformationTest.testWrongTagLogsInfoAndReturnsLocalMeta(DataServiceConnectionInformationTest.java:163)
Caused by: java.lang.ClassNotFoundException: org.json.simple.JSONObject
	at org.pentaho.di.trans.dataservice.client.DataServiceC
```

#### Unit test coverage change

These statistics help you identify how your changes have affected the coverage of the following files. If a file is not in this list, then its coverage was not affected by your changes. To get some help interpreting these metrics, please refer to [Jacoco's documentation](http://eclemma.org/jacoco/trunk/doc/counters.html).

###### org.pentaho.di.trans.dataservice.client.DataServiceConnectionInformation
 * Branch New Coverage: 20.00%
 * Complexity New Coverage: 53.33%
 * Instruction New Coverage: 30.99%
 * Line New Coverage: 36.84%
 * Method New Coverage: 80.00%

#### License header violations

> :heavy_exclamation_mark: Copyright year is not the current year on pdi-dataservice-client/src/main/java/org/pentaho/di/trans/dataservice/client/DataServiceConnectionInformation.java. Found 2016. Was expecting 2017

> :heavy_exclamation_mark: Copyright year is not the current year on pdi-dataservice-client/src/test/java/org/pentaho/di/trans/dataservice/client/DataServiceConnectionInformationTest.java. Found 2016. Was expecting 2017

--- a/.gitignore
+++ b/.gitignore
@@ -4,3 +4,4 @@ target/
 .settings/
 *.iml
 .idea/
+.DS_Store

--- a/pdi-dataservice-client/pom.xml
+++ b/pdi-dataservice-client/pom.xml
@@ -109,6 +109,12 @@
       <version>3.3.0-I20070605-0010</version>
       <scope>test</scope>
     </dependency>
+    <dependency>
+      <groupId>com.googlecode.json-simple</groupId>
+      <artifactId>json-simple</artifactId>
+      <version>1.1</version>
+      <scope>test</scope>
+    </dependency>
   </dependencies>
   <build>
     <finalName>${project.artifactId}</finalName>

--- a/pdi-dataservice-client/src/main/java/org/pentaho/di/trans/dataservice/client/DataServiceConnectionInformation.java
+++ b/pdi-dataservice-client/src/main/java/org/pentaho/di/trans/dataservice/client/DataServiceConnectionInformation.java
@@ -1,0 +1,118 @@
+/*! ******************************************************************************
+ *
+ * Pentaho Community Edition Project: data-refinery-pdi-plugin
+ *
+ * Copyright (C) 2002-2017 by Pentaho : http://www.pentaho.com
+ *
+ * *******************************************************************************
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ ********************************************************************************/
+
+package org.pentaho.di.trans.dataservice.client;
+
+import org.pentaho.di.core.ProvidesDatabaseConnectionInformation;
+import org.pentaho.di.core.database.DatabaseMeta;
+import org.pentaho.di.core.exception.KettleException;
+import org.pentaho.di.core.logging.LogChannelInterface;
+import org.pentaho.di.core.xml.XMLHandler;
+import org.pentaho.di.i18n.BaseMessages;
+import org.pentaho.di.repository.Repository;
+import org.w3c.dom.Document;
+import org.w3c.dom.Node;
+import org.w3c.dom.NodeList;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+
+public class DataServiceConnectionInformation implements ProvidesDatabaseConnectionInformation {
+  public static final String KETTLE_THIN = "KettleThin";
+  public static final String NATIVE = "native";
+  private String dataServiceName;
+  private Repository repository;
+  private LogChannelInterface log;
+  private static Class<?> PKG = DataServiceConnectionInformation.class; // for i18n purposes, needed by Translator2!!
+
+  public DataServiceConnectionInformation( String dataService, Repository repository, final LogChannelInterface log ) {
+    this.dataServiceName = dataService;
+    this.repository = repository;
+    this.log = log;
+  }
+
+  @Override public DatabaseMeta getDatabaseMeta() {
+    try {
+      return isDIRepository() ? getRepositoryDatabaseMeta() : getLocalDatabaseMeta();
+    } catch ( KettleException e ) {
+      log.logDebug( e.getMessage() );
+      return getLocalDatabaseMeta();
+    }
+  }
+
+  private boolean isDIRepository() {
+    try {
+      return repository != null && "PentahoEnterpriseRepository".equals( getRepositoryProperty( "id" ) );
+    } catch ( KettleException e ) {
+      log.logDebug( e.getMessage() );
+      return false;
+    }
+  }
+
+  private DatabaseMeta getRepositoryDatabaseMeta() throws KettleException {
+    String repositoryUrl = getRepositoryProperty( "repository_location_url" );
+    try {
+      URI uri = new URI( repositoryUrl );
+      DatabaseMeta databaseMeta =
+        new DatabaseMeta( dataServiceName, KETTLE_THIN, NATIVE, uri.getHost(), KETTLE_THIN,
+          Integer.toString( uri.getPort() ), repository.getUserInfo().getLogin(),
+          repository.getUserInfo().getPassword() );
+      databaseMeta.setDBName( uri.getPath().substring( 1 ) );
+      return databaseMeta;
+    } catch ( URISyntaxException e ) {
+      throw new KettleException( e );
+    }
+  }
+
+  private String getRepositoryProperty( final String nodeName ) throws KettleException {
+    final String xml = repository.getRepositoryMeta().getXML();
+    final Document doc = XMLHandler.loadXMLString( xml );
+    final NodeList childNodes = doc.getDocumentElement().getChildNodes();
+    for ( int i = 0; i < childNodes.getLength(); i++ ) {
+      final Node item = childNodes.item( i );
+      if ( nodeName.equals( item.getNodeName() ) ) {
+        return item.getTextContent();
+      }
+    }
+    throw new KettleException( BaseMessages.getString( PKG, "BuildModelJob.Info.ElementNotFound", nodeName ) );
+  }
+
+  private DatabaseMeta getLocalDatabaseMeta() {
+    DatabaseMeta databaseMeta =
+      new DatabaseMeta( dataServiceName, KETTLE_THIN, NATIVE, null, KETTLE_THIN, null, null, null );
+    databaseMeta.addExtraOption( KETTLE_THIN, "local", "true" );
+    return databaseMeta;
+  }
+
+  @Override public String getTableName() {
+    return dataServiceName;
+  }
+
+  @Override public String getSchemaName() {
+    return "Kettle";
+  }
+
+  @Override public String getMissingDatabaseConnectionInformationMessage() {
+    // Use default connection missing message
+    return null;
+  }
+}

--- a/pdi-dataservice-client/src/main/resources/org/pentaho/di/trans/dataservice/client/messages/messages_en_US.properties
+++ b/pdi-dataservice-client/src/main/resources/org/pentaho/di/trans/dataservice/client/messages/messages_en_US.properties
@@ -1,0 +1,1 @@
+BuildModelJob.Info.ElementNotFound="Element {0} not found in Repository Meta XML"

--- a/pdi-dataservice-client/src/test/java/org/pentaho/di/trans/dataservice/client/DataServiceClientPluginTest.java
+++ b/pdi-dataservice-client/src/test/java/org/pentaho/di/trans/dataservice/client/DataServiceClientPluginTest.java
@@ -2,7 +2,7 @@
  *
  * Pentaho Data Integration
  *
- * Copyright (C) 2002-2016 by Pentaho : http://www.pentaho.com
+ * Copyright (C) 2002-2017 by Pentaho : http://www.pentaho.com
  *
  *******************************************************************************
  *

--- a/pdi-dataservice-client/src/test/java/org/pentaho/di/trans/dataservice/client/DataServiceConnectionInformationTest.java
+++ b/pdi-dataservice-client/src/test/java/org/pentaho/di/trans/dataservice/client/DataServiceConnectionInformationTest.java
@@ -1,0 +1,211 @@
+/*! ******************************************************************************
+ *
+ * Pentaho Community Edition Project: data-refinery-pdi-plugin
+ *
+ * Copyright (C) 2002-2017 by Pentaho : http://www.pentaho.com
+ *
+ * *******************************************************************************
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ ********************************************************************************/
+
+package org.pentaho.di.trans.dataservice.client;
+
+import org.junit.After;
+import org.junit.AfterClass;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.mockito.Mockito;
+import org.mockito.invocation.InvocationOnMock;
+import org.mockito.stubbing.Answer;
+import org.pentaho.di.core.KettleEnvironment;
+import org.pentaho.di.core.database.DatabaseInterface;
+import org.pentaho.di.core.database.DatabaseMeta;
+import org.pentaho.di.core.logging.LogChannelInterface;
+import org.pentaho.di.core.plugins.DatabasePluginType;
+import org.pentaho.di.core.plugins.PluginInterface;
+import org.pentaho.di.core.plugins.PluginRegistry;
+import org.pentaho.di.core.plugins.PluginTypeInterface;
+import org.pentaho.di.repository.IUser;
+import org.pentaho.di.repository.Repository;
+import org.pentaho.di.repository.RepositoryMeta;
+import org.pentaho.di.trans.dataservice.client.DataServiceClientPlugin;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNull;
+import static org.mockito.Matchers.anyString;
+import static org.mockito.Mockito.*;
+
+public class DataServiceConnectionInformationTest {
+  private static PluginInterface mockDbPlugin;
+  private LogChannelInterface log;
+
+  @Before
+  public void setUp() throws Exception {
+    log = mock( LogChannelInterface.class );
+  }
+
+  @After
+  public void tearDown() throws Exception {
+    verifyNoMoreInteractions( log );
+  }
+
+  @BeforeClass
+  public static void setUpBeforeClass() throws Exception {
+    if ( !KettleEnvironment.isInitialized() ) {
+      KettleEnvironment.init();
+    }
+    Map<Class<?>, String> dbMap = new HashMap<Class<?>, String>();
+    dbMap.put( DatabaseInterface.class, DataServiceClientPlugin.class.getName() );
+    PluginRegistry preg = PluginRegistry.getInstance();
+    mockDbPlugin = mock( PluginInterface.class );
+    when( mockDbPlugin.matches( anyString() ) ).thenReturn( true );
+    when( mockDbPlugin.isNativePlugin() ).thenReturn( true );
+    when( mockDbPlugin.getMainType() ).thenAnswer( new Answer<Class<?>>() {
+      @Override
+      public Class<?> answer( InvocationOnMock invocation ) throws Throwable {
+        return DatabaseInterface.class;
+      }
+    } );
+
+    when( mockDbPlugin.getPluginType() ).thenAnswer( new Answer<Class<? extends PluginTypeInterface>>() {
+      @Override
+      public Class<? extends PluginTypeInterface> answer( InvocationOnMock invocation ) throws Throwable {
+        return DatabasePluginType.class;
+      }
+    } );
+
+    when( mockDbPlugin.getIds() ).thenReturn( new String[] { "KettleThin" } );
+    when( mockDbPlugin.getName() ).thenReturn( "KettleThin" );
+    when( mockDbPlugin.getClassMap() ).thenReturn( dbMap );
+
+    preg.registerPlugin( DatabasePluginType.class, mockDbPlugin );
+  }
+
+  @AfterClass
+  public static void tearDownClass() throws Exception {
+    PluginRegistry.getInstance().removePlugin( DatabasePluginType.class, mockDbPlugin );
+  }
+
+  @Test
+  public void testNullRepositoriesReturnsLocalMeta() throws Exception {
+    String dataServiceName = "myDataService";
+    DataServiceConnectionInformation connectInfo = new DataServiceConnectionInformation( dataServiceName, null, log );
+    assertLocalConnection( dataServiceName, connectInfo );
+  }
+
+  @Test
+  public void testDIRepositoriesReturnsMetaToSameHost() throws Exception {
+    String dataServiceName = "myDataService";
+    final Repository repository = expectRepository(
+      "  <repository>"
+        + "  <id>PentahoEnterpriseRepository</id>\n"
+        + "  <name>local</name>\n"
+        + "  <description>a</description>\n"
+        + "  <repository_location_url>http&#x3a;&#x2f;&#x2f;farfaraway&#x3a;12345&#x2f;pentaho</repository_location_url>\n"
+        + "  <version_comment_mandatory>N</version_comment_mandatory>\n"
+        + "</repository>"
+    );
+    final IUser user = mock( IUser.class );
+    when( repository.getUserInfo() ).thenReturn( user );
+    when( user.getLogin() ).thenReturn( "open" );
+    when( user.getPassword() ).thenReturn( "sesame" );
+    DataServiceConnectionInformation connectInfo = new DataServiceConnectionInformation( dataServiceName, repository,
+      log );
+    assertEquals( dataServiceName, connectInfo.getTableName() );
+    assertEquals( "Kettle", connectInfo.getSchemaName() );
+    DatabaseMeta databaseMeta = connectInfo.getDatabaseMeta();
+    assertEquals( "org.pentaho.di.trans.dataservice.jdbc.ThinDriver", databaseMeta.getDriverClass() );
+    assertEquals( "pentaho", databaseMeta.getDatabaseName() );
+    assertEquals( dataServiceName, databaseMeta.getName() );
+    assertNull( databaseMeta.getExtraOptions().get( "KettleThin.local" ) );
+    assertEquals( "farfaraway", databaseMeta.getHostname() );
+    assertEquals( "12345", databaseMeta.getDatabasePortNumberString() );
+    assertEquals( "open", databaseMeta.getUsername() );
+    assertEquals( "sesame", databaseMeta.getPassword() );
+  }
+
+  @Test
+  public void testOtherRepositoryTypesReturnLocalMeta() throws Exception {
+    String dataServiceName = "yourDataService";
+    final Repository repository = expectRepository(
+      "<repository>"
+      + "  <id>KettleFileRepository</id>\n"
+      + "  <name>local</name>\n"
+      + "  <description>a</description>\n"
+      + "  <repository_location_url>http&#x3a;&#x2f;&#x2f;farfaraway&#x3a;12345&#x2f;pentaho</repository_location_url>\n"
+      + "  <version_comment_mandatory>N</version_comment_mandatory>\n"
+      + "</repository>" );
+    DataServiceConnectionInformation connectInfo = new DataServiceConnectionInformation( dataServiceName, repository,
+      log );
+    assertLocalConnection( dataServiceName, connectInfo );
+  }
+
+  @Test
+  public void testWrongTagLogsInfoAndReturnsLocalMeta() throws Exception {
+    String dataServiceName = "otherDataService";
+    Repository repository = expectRepository(
+      "<repository>"
+      + "  <id>PentahoEnterpriseRepository</id>\n"
+      + "  <name>local</name>\n"
+      + "  <description>a</description>\n"
+      + "  <repository_location>http&#x3a;&#x2f;&#x2f;farfaraway&#x3a;12345&#x2f;pentaho</repository_location>\n"
+      + "  <version_comment_mandatory>N</version_comment_mandatory>\n"
+      + "</repository>" );
+    DataServiceConnectionInformation connectInfo =
+      new DataServiceConnectionInformation( dataServiceName, repository, log );
+    assertLocalConnection( dataServiceName, connectInfo );
+    verify( log ).logDebug( "\n\"Element repository_location_url not found in Repository Meta XML\"\n" );
+  }
+
+  @Test
+  public void testBadXmlLogsInfoAndReturnsLocalMeta() throws Exception {
+    String dataServiceName = "otherDataService";
+    Repository repository = expectRepository(
+      "<reposito></repository>" );
+    DataServiceConnectionInformation connectInfo =
+      new DataServiceConnectionInformation( dataServiceName, repository, log );
+    assertLocalConnection( dataServiceName, connectInfo );
+    verify( log ).logDebug( Mockito.contains( "Error reading information from XML string" ) );
+  }
+
+  @Test
+  public void testNoSpecificMessageForMissingParameters() throws Exception {
+    DataServiceConnectionInformation connectionInformation =
+      new DataServiceConnectionInformation( "", null, log );
+    assertNull( connectionInformation.getMissingDatabaseConnectionInformationMessage() );
+  }
+
+  private Repository expectRepository( final String xml ) {
+    final Repository repository = mock( Repository.class );
+    final RepositoryMeta repositoryMeta = mock( RepositoryMeta.class );
+    when( repository.getRepositoryMeta() ).thenReturn( repositoryMeta );
+    when( repositoryMeta.getXML() ).thenReturn( xml );
+    return repository;
+  }
+
+  private void assertLocalConnection( final String dataServiceName, final DataServiceConnectionInformation connectInfo ) {
+    assertEquals( dataServiceName, connectInfo.getTableName() );
+    assertEquals( "Kettle", connectInfo.getSchemaName() );
+    DatabaseMeta databaseMeta = connectInfo.getDatabaseMeta();
+    assertEquals( "org.pentaho.di.trans.dataservice.jdbc.ThinDriver", databaseMeta.getDriverClass() );
+    assertEquals( "KettleThin", databaseMeta.getDatabaseName() );
+    assertEquals( dataServiceName, databaseMeta.getName() );
+    assertEquals( "true", databaseMeta.getExtraOptions().get( "KettleThin.local" ) );
+  }
+}

